test/001_highsnr_stdio_audio/test_callback_rx.py

```diff
@@ -0,0 +1,178 @@
+TIMEOUT = args.TIMEOUT
+
+# AUDIO PARAMETERS
+AUDIO_FRAMES_PER_BUFFER = 2400*2  # <- consider increasing if you get nread_exceptions > 0
+MODEM_SAMPLE_RATE = codec2.api.FREEDV_FS_8000
+AUDIO_SAMPLE_RATE_RX = 48000
+
+# make sure our resampler will work
+assert (AUDIO_SAMPLE_RATE_RX / MODEM_SAMPLE_RATE) == codec2.api.FDMDV_OS_48
+
+
+# ------------------------------------------------ PYAUDIO CALLBACK
+def callback(data_in48k, frame_count, time_info, status):
+    x = np.frombuffer(data_in48k, dtype=np.int16)
+    x.tofile(frx)    
+    x = resampler.resample48_to_8(x)
+    audio_buffer.push(x)    
+    return (None, pyaudio.paContinue)
+
+
+# check if we want to use an audio device then do an pyaudio init
+if AUDIO_INPUT_DEVICE != -1: 
+    p = pyaudio.PyAudio()
+    # auto search for loopback devices
+    if AUDIO_INPUT_DEVICE == -2:
+        loopback_list = []
+        for dev in range(0,p.get_device_count()):
+            if 'Loopback: PCM' in p.get_device_info_by_index(dev)["name"]:
+                loopback_list.append(dev)
+        if len(loopback_list) >= 2:
+            AUDIO_INPUT_DEVICE = loopback_list[0] #0  = RX   1 = TX
+            print(f"loopback_list rx: {loopback_list}", file=sys.stderr)
+        else:
+            quit()
+            
+    print(f"AUDIO INPUT DEVICE: {AUDIO_INPUT_DEVICE} DEVICE: {p.get_device_info_by_index(AUDIO_INPUT_DEVICE)['name']}  \
+            AUDIO SAMPLE RATE: {AUDIO_SAMPLE_RATE_RX}", file=sys.stderr)
+    
+    stream_rx = p.open(format=pyaudio.paInt16, 
+                            channels=1,
+                            rate=AUDIO_SAMPLE_RATE_RX,
+                            frames_per_buffer=AUDIO_FRAMES_PER_BUFFER,
+                            input=True,
+                            output=False,
+                            input_device_index=AUDIO_INPUT_DEVICE,
+                            stream_callback=callback
+                            ) 
+    try:                        
+        print(f"starting pyaudio callback", file=sys.stderr)
+        stream_rx.start_stream()
+    except Exception as e:
+        print(f"pyAudio error: {e}", file=sys.stderr)  
+# ----------------------------------------------------------------
+              
+# DATA CHANNEL INITIALISATION
+
+# open codec2 instance        
+freedv = cast(codec2.api.freedv_open(MODE), c_void_p)
+
+# get number of bytes per frame for mode
+bytes_per_frame = int(codec2.api.freedv_get_bits_per_modem_frame(freedv)/8)
+payload_bytes_per_frame = bytes_per_frame -2
+
+n_max_modem_samples = codec2.api.freedv_get_n_max_modem_samples(freedv)     
+bytes_out = create_string_buffer(bytes_per_frame * 2)
+
+codec2.api.freedv_set_frames_per_burst(freedv,N_FRAMES_PER_BURST)
+
+total_n_bytes = 0
+rx_total_frames = 0
+rx_frames = 0
+rx_bursts = 0
+rx_errors = 0
+nread_exceptions = 0
+timeout = time.time() + TIMEOUT
+receive = True
+audio_buffer = codec2.audio_buffer(AUDIO_FRAMES_PER_BUFFER*2)
+resampler = codec2.resampler()
+
```

Comment:
@DJ2LS I think this needs to be created before the stream is started, to set up the state of the resampler.

Reply:
I changed the order of this part also this afternoon and now again, but the resampler still crashes. I would'nt be surprised if the problem is pyAudio's callback data which differs from the normal blocking mode. I had some curious behavior when using the callback function some weeks ago for writing data to the audio stream :thinking: But I'm not sure if we really need the callback mode. The next days will maybe clarify this, I think.

Reply:
Yes I have been going down the same path ... :thinking: